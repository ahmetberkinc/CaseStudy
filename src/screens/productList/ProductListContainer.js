import React, {useEffect, useState, useRef} from 'react';
import ProductListView from './ProductListView';
import {
  filterProductList,
  getAllProducts,
  getProductList,
} from '../../services/productApi';

const ProductListContainer = () => {
  const [displayedProducts, setDisplayedProducts] = useState([]);
  const [allProducts, setAllProducts] = useState([]);

  //Display loading icon with footer element of FlatList
  const [isLoading, setIsLoading] = useState(false);

  //Responsible for detecting is product last page of array.
  const [hasMoretoLoad, setHasMoretoLoad] = useState(true);

  const [selectedFilterOption, setSelectedFilterOption] = useState('ALL');

  const page = useRef(1);

  useEffect(() => {
    getProducts();
    getEntireProducts();
  }, []);

  //This api call responsible for displaying all available filter options
  function getEntireProducts() {
    getAllProducts().then(result => {
      setAllProducts(result);
    });
  }

  //Get products with limit=12&pagination when page is render
  function getProducts() {
    setIsLoading(true);
    getProductList(page.current, selectedFilterOption)
      .then(result => {
        setDisplayedProducts([...displayedProducts, ...result]);
        page.current += 1;
        setIsLoading(false);
        if (result.length < 12) {
          setHasMoretoLoad(false);
        }
      })
      .catch(setIsLoading(false));
  }

  //Filter products with user brand filter option
  function filterProducts(keyToFilterBy) {
    setIsLoading(true);
    page.current = 1;
    filterProductList(page.current, keyToFilterBy)
      .then(result => {
        setDisplayedProducts(result);
        setIsLoading(false);
        if (result.length < 12) {
          setHasMoretoLoad(false);
        }
      })
      .catch(setIsLoading(false));
  }

  return (
    <ProductListView
      displayedProducts={displayedProducts}
      allProducts={allProducts}
      onEndReached={getProducts}
      isLoading={isLoading}
      hasMoretoLoad={hasMoretoLoad}
      setSelectedFilterOption={setSelectedFilterOption}
      selectedFilterOption={selectedFilterOption}
      onFilterSelection={filterKey => filterProducts(filterKey)}
    />
  );
};

export default ProductListContainer;
