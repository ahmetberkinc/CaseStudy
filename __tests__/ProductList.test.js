import React from 'react';
import renderer from 'react-test-renderer';
import ProductListContainer from '../src/screens/productList/ProductListContainer';

test('renders correctly', () => {
  const tree = renderer.create(<ProductListContainer />).toJSON();
  expect(tree).toMatchSnapshot();
});
