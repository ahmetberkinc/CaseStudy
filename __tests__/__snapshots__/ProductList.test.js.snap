// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "flex": 1,
      "marginHorizontal": 16,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#D9D9D9",
        "borderColor": "#D9D9D9",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 40,
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 30,
          },
          undefined,
          {
            "fontFamily": "EvilIcons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      ï„¹
    </Text>
    <TextInput
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Search"
      style={
        {
          "flex": 1,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
      value=""
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        {
          "color": "#000000",
          "flex": 1,
          "fontSize": 18,
          "fontWeight": "500",
        }
      }
    >
      Filters:
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "flex-end",
          "backgroundColor": "#D9D9D9",
          "flexDirection": "row",
          "height": 36,
          "justifyContent": "center",
          "marginVertical": 14,
          "opacity": 1,
          "width": 158,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 14,
          }
        }
      >
        Select Filter
      </Text>
      <Modal
        animationType="slide"
        hardwareAccelerated={false}
        transparent={true}
        visible={false}
      />
    </View>
  </View>
  <RCTScrollView
    ListFooterComponent={[Function]}
    contentContainerStyle={
      {
        "paddingBottom": 32,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 10,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
